### PR TITLE
Add query grid

### PR DESCRIPTION
**Pull request by tsauerwein**
Closes https://github.com/camptocamp/c2cgeoportal/issues/1655

Still to do:

* [x] Support HTML/images in table cells
* [x] Do separate query to get count before loading features
* [x] CSV export

### CODE REVIEW

**Comment by sbrunner** [comment]
Visible some where?

**Comment by tsauerwein** [comment]
> Visible some where?

Here you go: https://tsauerwein.github.io/ngeo/querygrid/examples/contribs/gmf/apps/desktop_alt/

**Comment by sbrunner** [comment]
The demo looks good :-)

Some remarques:
 * I don't see the count
 * The export CSV button don't works

**Comment by tsauerwein** [comment]
> I don't see the count

Which count? Behind the source title in the "nav pills" you should see the number of features for each source.
ngeo-grid


> The export CSV button doesn't work

Could you be more specific? What does not work? Do you get an error?

**Comment by ger-benjamin** [comment]
Looks great ! I'll take a look in the code later.

Can you take in account my changes in this PR (you just need to add the `ngeo-map-query-autoclear` and `ngeo-bbox-query-autoclear` attributes in the respective directives declarations in the desktop_alt/index.html file.): See my change in the desktop/index.html file https://github.com/camptocamp/ngeo/pull/1706/files

**Comment by sbrunner** [comment]
> Could you be more specific? What does not work? Do you get an error?

no request , no error...

**Comment by tsauerwein** [comment]
What browser are you using?

**Comment by sbrunner** [comment]
> I don't see the count...

Effectively, I just search it on the status bar :-)

**Comment by sbrunner** [comment]
> What browser are you using?

Firefox 48

**Comment by tsauerwein** [comment]
> Firefox 48

Ok, thanks. I found out what was wrong.

File: src/filter/trusthtml.js [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+ * @param {angular.$sce} $sce Angular sce service.
+ * @ngname ngeoTrustHtml
+ */
+ngeo.trustHtmlFilter = function($sce) {
+  return function(input) {
+    return $sce.trustAsHtml(input);
+  };
+};
+
+ngeo.module.filter('ngeoTrustHtml', ngeo.trustHtmlFilter);
```
Comment (sbrunner) [inline src/filter/trusthtml.js]: I don't understand what's for but filters are actually in `src/service/filters.js`...

Comment (tsauerwein) [inline src/filter/trusthtml.js]: Thanks, I did not expect the filters to be in the service package.

**Comment by tsauerwein** [comment]
> Can you take in account my changes in this PR ...

What are your changes supposed to do? Currently the result grid stays open if you open the print panel.

**Comment by ger-benjamin** [comment]
>Can you take in account my changes in this PR ...
>>What are your changes supposed to do? Currently the result grid stays open if you open the print panel.

With my change it's impossible to do another query when the print panel is open (that's a specification).

**Comment by tsauerwein** [comment]
> With my change it's impossible to do another query when the print panel is open (that's a specification).

Ok, done.

File: contribs/gmf/src/directives/displayquerygrid.js [hunk 2]
```diff
@@ -0,0 +1,788 @@
...
+            this.onSelectionChanged_();
+          }
+        }.bind(this));
+  }
+  this.updateFeatures_(gridSource);
+};
+
+
+/**
+ * Called when the row selection has changed.
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/displayquerygrid.js]: Missing tag `@private`

File: contribs/gmf/src/directives/displayquerygrid.js [hunk 3]
```diff
@@ -0,0 +1,788 @@
...
+  }
+
+  var gridSource = this.gridSources['' + this.selectedTab];
+  this.updateFeatures_(gridSource);
+};
+
+
+/**
+ * @param {gmfx.GridSource} gridSource Grid source
+ * @export
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/displayquerygrid.js]: Missmatching tag `@export` and function name with underscore

**Comment by ger-benjamin** [comment]
Please, add documentation for all function with an `@export` tag

File: src/directives/grid.js [hunk 4]
```diff
@@ -0,0 +1,397 @@
...
+ */
+ngeo.GridConfig.prototype.getSelectedRows = function() {
+  return this.data.filter(function(row) {
+    return this.isRowSelected(row);
+  }.bind(this));
+};
+
+
+/**
+ * @param {Object} attributes An entry/row.
```
Comment (ger-benjamin) [inline src/directives/grid.js]: `@export` (and description) ?

File: src/directives/grid.js [hunk 5]
```diff
@@ -0,0 +1,397 @@
...
+ */
+ngeo.GridConfig.prototype.selectRow = function(attributes) {
+  var uid = ngeo.GridConfig.getRowUid(attributes);
+  this.selectedRows[uid] = attributes;
+};
+
+
+/**
+ * @param {Object} attributes An entry/row.
```
Comment (ger-benjamin) [inline src/directives/grid.js]: `@export` (and description) ?

File: src/services/csvdownload.js [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  }.bind(this));
+
+  return header + dataRows.join('');
+};
+
+
+/**
+ * @param {Array.<?>} values Values.
+ * @return {string} CSV row.
+ * @export
```
Comment (ger-benjamin) [inline src/services/csvdownload.js]: Missmatch between `@export` and function name (underscore)

File: contribs/gmf/src/directives/displayquerygrid.js [hunk 7]
```diff
@@ -0,0 +1,788 @@
...
+    restrict: 'E',
+    scope: {
+      'active': '=gmfDisplayquerygridActive',
+      'featuresStyleFn': '&gmfDisplayquerygridFeaturesstyle',
+      'selectedFeatureStyleFn': '&gmfDisplayquerygridSourceselectedfeaturestyle',
+      'getMapFn': '&gmfDisplayquerygridMap',
+      'removeEmptyColumnsFn': '&?gmfDisplayquerygridRemoveemptycolumns',
+      'maxResultsFn': '&?gmfDisplayquerygridMaxresults',
+      'maxRecenterZoomFn': '&?gmfDisplayquerygridMaxrecenterzoom',
+      'mergeTabsFn': '&gmfDisplayquerygridMergetabs'
```
Comment (ger-benjamin) [inline contribs/gmf/src/directives/displayquerygrid.js]: That should not be `&?gmfDisplayquerygridMergetabs` ? (bound with `&?` instead of `&`) ?

**Comment by ger-benjamin** [comment]
> Support HTML/images in table cells

Do you have an example (name of a layer) with this feature ?

**Comment by ger-benjamin** [comment]
That's a little bit a stupid case but if you select a feature without data, I think that the row must have a minimum size (even without text).

Try it on this polygon (more interesting with multiple rows) : https://tsauerwein.github.io/ngeo/querygrid/examples/contribs/gmf/apps/desktop_alt/?baselayer_ref=map&lang=en&map_x=540475&map_y=156520&map_zoom=5&tree_groups=Edit

selection_063

**Comment by ger-benjamin** [comment]
Except my comments, it looks really good :-)
Thanks !

**Comment by tsauerwein** [comment]
> Please, add documentation for all functions with an @export tag
> (Sometimes that seems useless in the code but that's not in the apidoc)

I didn't know that the documentation for `@returns` is not shown in the apidocs. I will change it.

**Comment by tsauerwein** [comment]
>  > Support HTML/images in table cells
> 
> Do you have an example (name of a layer) with this feature ?

"Fonctions OSM mixed > Ouverture OSM". That layer contains HTML/JS as attribute. An old cgxp function is used though.

**Comment by ger-benjamin** [comment]
Thanks,

Mhhh in the last commit, you've used the tag `@public` and not the tag `@export`. It's an error no ?

**Comment by tsauerwein** [comment]
> Mhhh in the last commit, you've used the tag @public and not the tag @export. It's an error no ?

These two methods are not used in a partial.

**Comment by ger-benjamin** [comment]
Okay, so if the tag `@public` cause no exception by the code checks it's a good idea. I don't like when there is no tag. @public is clear.

**Comment by ger-benjamin** [comment]
For me it ready to be merged. If we find issues, they can be resolved in another (smaller) PR

**Comment by tsauerwein** [comment]
> That's a little bit a stupid case but if you select a feature without data, I think that the row must have a minimum size (even if the row as no text/data).

There is this property `removeemptycolumns` which is set to `true`. In your case all columns were empty. With my latest commit, the grid is not shown if there are no columns left.

**Comment by tsauerwein** [comment]
Thanks for your review, I will merge when Travis passes.

**Pull request merged**